Fix "list_type_missing" API violations in meta/v1

thockin opened the pull request:
This assumes that any such field is atomic, except:
  * OwnerReferences: because it has a `+patchStrategy=merge`, but it probably needs a `+listMapKey=...` ?
  * Finalizers: because it hs a `+patchStrategy=merge`, but is a primitive type (string).

An alternative approach could be just to turn off the API warnings for these fields, but it felt more correct to declare the semantics.

If this flies I can do others.

@apelisse to tell me if I totally missed the mark on this assumption and to advise on the above.
@liggitt FYI
@deads2k For consult

/kind cleanup
```release-note
NONE
```

thockin committed:
Fix "list_type_missing" API violations in meta/v1

This assumes that any such field is atomic, except:
  * OwnerReferences: because it has a `+patchStrategy=merge`, but it
    probably needs a `+listMapKey=...` ?
  * Finalizers: because it hs a `+patchStrategy=merge`, but is a
    primitive type (string).
  * []byte fields, which should not be failing this anyway (fixed
    subsequently).

An alternative approach could be just to turn off the API warnings for
these fields, but it felt more correct to declare the semantics.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/types.go:
```diff
@@ -102,6 +102,7 @@ type RawExtension struct {
 	// Raw is the underlying serialization of this object.
 	//
 	// TODO: Determine how to detect ContentType and ContentEncoding of 'Raw' data.
+	// +listType=atomic
```
these look ~fine for now, but we could also fix the linter to not complain about `[]byte` "lists". cc @apelisse

thockin replied:
https://github.com/kubernetes/kube-openapi/pull/437

thockin replied:
I will back these out from this PR, but as a new commit for readability

alexzielenski commented:
/triage accepted

thockin commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -236,6 +236,7 @@ type ObjectMeta struct {
 	// +optional
 	// +patchMergeKey=uid
 	// +patchStrategy=merge
+	// +listType=map
```
I wondered about it.  In next push.

thockin commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -268,6 +270,7 @@ type ObjectMeta struct {
 	// workflow used when modifying the object.
 	//
 	// +optional
+	// +listType=atomic
```
Would `+listType=None` make sense? (or something like that) ?

liggitt replied:
in theory, these could still be included in an API of some type, so defining them as listType=atomic seems ok

thockin committed:
Fix list_type_missing in apiextensions-apiserver

thockin committed:
Fix list_type_missing in sample-apiserver

thockin committed:
Fix list_type_missing in api/admissionregistration

thockin committed:
Fix list_type_missing in api/apps

thockin committed:
Fix list_type_missing in api/authentication

thockin committed:
Fix list_type_missing in api/authorization

thockin committed:
Fix list_type_missing in api/autoscaling

thockin committed:
Fix list_type_missing in api/core

thockin commented:
It would be nice if there was a doc which indicated the implicit defaults in each case.  E.g. if I set CSA "merge" on a `[]string` what is the correct SSA?

thockin committed:
Re-vendor k8s.io/kube-openapi

./hack/pin-dependency.sh k8s.io/kube-openapi 778a5567bc1edaed92a4de9c07f90199c67953fa

./hack/update-vendor.sh

thockin committed:
Re-run codegen with new kube-openapi

thockin commented:
Now includes re-vendor of kube-openapi with no `[]byte` warnings.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1218,6 +1238,7 @@ type LabelSelector struct {
 	MatchLabels map[string]string `json:"matchLabels,omitempty" protobuf:"bytes,1,rep,name=matchLabels"`
 	// matchExpressions is a list of label selector requirements. The requirements are ANDed.
 	// +optional
+	// +listType=atomic
```
@apelisse should the linter stop checking listType within a struct labeled as `+structType=atomic`?

thockin replied:
I did not know `+structType=atomic` was a thing!  None of this seems to be documented.  I'll try removing that and see if it complains.

thockin replied:
Nope, it still complains:

```
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,LabelSelector,MatchExpressions
```

thockin committed:
Fix list_type_missing in api/extensions

thockin committed:
Fix list_type_missing in api/imagepolicy

thockin committed:
Fix list_type_missing in api/networking

thockin committed:
Fix list_type_missing in api/rbac

thockin committed:
Fix list_type_missing in api/storage

thockin commented:
/retest

liggitt commented:
changes lgtm, let me know when this is no longer WIP and you want a final check

thockin commented:
I was going to keep chipping at it until I get them all, since we can't merge yet anyway

thockin committed:
Fix list_type_missing in k8s.io/apiserver

thockin commented:
Opinion wanted: I'm now into things like "k8s.io/kube-controller-manager/config/v1alpha1/types.go" - these are not really APIs tha need "apply" logic, right?  Maybe the warning tool should not be considering them?

thockin committed:
Fix list_type_missing in k8s.io/metrics

liggitt commented:
> Opinion wanted: I'm now into things like "k8s.io/kube-controller-manager/config/v1alpha1/types.go" - these are not really APIs tha need "apply" logic, right? Maybe the warning tool should not be considering them? @liggitt @apelisse

There's bits in the warning tool (like field names matching serialized field names) that are useful even for config types. Agree the apply warnings don't really... apply... there.

apelisse commented:
> There's bits in the warning tool (like field names matching serialized field names) that are useful even for config types. Agree the apply warnings don't really... apply... there.

;-)

What's the proper way to go then? A header to disable the warning? How do we know these types are "different"?

liggitt commented:
(unit test flake is https://github.com/kubernetes/kubernetes/issues/121926)

thockin commented:
/retest

pastequo commented:
Should this PR be back-ported in 1.29 and others ? It might be useful for Customer Resources that includes native types into their definition, to have the full granularity for managedfields

https://kubernetes.slack.com/archives/C0123CNN8F3/p1705594643726679

dprotaso commented:
Brought this up at sig-api-machinery this week and folks were ok with a cherry-picking this back ([notes](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#heading=h.9p525fj57pix), [video](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#heading=h.9p525fj57pix)) to a released version of K8s - in order to fix SSA

/cherrypick release-1.29

liggitt commented:
> Brought this up at sig-api-machinery this week and folks were ok with a cherry-picking this back ([notes](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#heading=h.9p525fj57pix), [video](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#heading=h.9p525fj57pix)) to a released version of K8s - in order to fix SSA

What specifically would this fix in SSA? the only two fields that actually changed anything effectively in this PR were ownerReferences and finalizers, right? neither of those were the ones mentioned in the linked slack thread

Backporting API behavior to a minor doesn't seem like a great idea to me. Letting the change take effect in the next release seems more reasonable.

dprotaso commented:
> What specifically would this fix in SSA? the only two fields that actually changed anything effectively in this PR were ownerReferences and finalizers, right? neither of those were the ones mentioned in the linked slack thread

Various fields in the PodSpec now have the proper listType and listMapKey annotations - thus for those that embed this type into their own CRDs they'll have the proper OpenAPI schema when generating these CRD definitions.

dprotaso commented:
> Backporting API behavior to a minor doesn't seem like a great idea to me. Letting the change take effect in the next release seems more reasonable.

Yeah that's fair - seems like the cherry-picker plugin isn't enabled anymore ;)

liggitt commented:
> Various fields in the PodSpec now have the proper listType and listMapKey annotations - thus for those that embed this type into their own CRDs they'll have the proper OpenAPI schema when generating these CRD definitions.

Ah. Letting folks bump to k8s.io/api v0.30.0 when it's released would let them regen with annotations, right? I'd probably opt for that.

> seems like the cherry-picker plugin isn't enabled anymore ;)

it was never enabled for kubernetes/kubernetes, fyi
